pydantic/dataclasses.py

```diff
@@ -188,6 +188,15 @@ def _process_class(
         cls.__name__, __config__=config, __module__=_cls.__module__, __validators__=validators, **field_definitions
     )
 
+    if getattr(config, 'extra', Extra.forbid) is not Extra.forbid:
+        def allow_extra_init(self, *args, **kwargs):
+            self.__original_init__(*args, **{
+                k: v for k, v in kwargs.items() if k in get_type_hints(type(self))
+            })
+
+        cls.__original_init__ = cls.__init__
+        cls.__init__ = allow_extra_init
```

Comment:
@PrettyWood I remember we had lots of problems with overriding `__init__` at some points.

Do you think this will break anything?

What about conflicts with #2557, will it just be code conflicts, or does this approach need to be rethought?

Reply:
I'm sorry @DetachHead but I felt like it was way easier to add it in https://github.com/samuelcolvin/pydantic/pull/2557/commits/593c9bb4abba752af31b16ecd229ee645331fe66 rather than handling conflicts
@samuelcolvin IMO if the commit above is good we can close this PR

Reply:
@PrettyWood looks good. though i can't find the branch that commit is on? is there a PR for it?